pkcs7/benchmark-streaming-envelop.c

```diff
@@ -0,0 +1,412 @@
+            byte tmpBuffer[1000];
+            int j;
+
+            for (j = 0; j < sz; j++) {
+                tmpBuffer[j] = rand() % 256;
+            }
+            sz = fwrite(tmpBuffer, 1, sz, f);
+            if (sz <= 0) {
+                printf("Failed to write to content file\n");
+                ret = -1;
+                break;
+            }
+            i += sz;
+        }
+        fclose(f);
+    }
+    return ret;
+}
+
+typedef struct BENCHMARK_IO {
+    FILE* in;
+    FILE* out;
+    byte  buf[TEST_STREAM_CHUNK_SIZE];
+} BENCHMARK_IO;
+
+
+/* Callback function to allow wolfSSL to read from a stream of input when
+ * working with the PKCS7 content. The byte pointer content is handled by the
+ * code managing the callback, meaning that malloc'ing/free'ing any memory
+ * should be done here, wolfSSL will not try to free the pointer given.
+ *
+ * Expected to return the number of bytes that the buffer pointed to contains */
+static int GetContentCB(PKCS7* pkcs7, byte** content, void* ctx)
+{
+    int ret = 0;
+    BENCHMARK_IO* io = (BENCHMARK_IO*)ctx;
+
+    if (io != NULL) {
+        ret = fread(io->buf, 1, TEST_STREAM_CHUNK_SIZE, io->in);
+        if (ret > 0) {
+            *content = io->buf;
+        }
+    }
+
+    (void)pkcs7;
+    return ret;
+}
+
+
+/* Callback function to output the PKCS7 bundle as it is created. The buffer
+ * 'output' contains the current data to be written out and 'outputSz' is the
+ * number of bytes in the 'output' buffer.
+ *
+ * Expected to return 0 on success.
+ */
+static int StreamOutputCB(PKCS7* pkcs7, const byte* output, word32 outputSz,
+    void* ctx)
+{
+    int ret = 0;
+    BENCHMARK_IO* io = (BENCHMARK_IO*)ctx;
+
+    if (io != NULL) {
+        ret = fwrite(output, 1, outputSz, io->out);
+        if (ret < 0) {
+            printf("stream output write failed\n");
+            ret = -1;
+        }
+    }
+
+    (void)pkcs7;
+    return 0;
+}
+
+
+static int EncodePKCS7Bundle(double contentSz, WC_RNG* rng)
+{
+    wc_PKCS7* pkcs7;
+    double per;
+    int ret = 0;
+    BENCHMARK_IO io;
+    byte aes256Key[] = {
+        0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08,
+        0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08,
+        0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08,
+        0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08
+    };
+
+    printf("Creating an encoded bundle ... ");
+    TimeLogStart();
+
+    pkcs7 = wc_PKCS7_New(NULL, 0);
+    if (pkcs7 == NULL) {
+        printf("Failed to create PKCS7 struct\n");
+        ret = MEMORY_E;
+    }
+
+    if (ret == 0) {
+        ret = wc_PKCS7_InitWithCert(pkcs7, (byte*)client_cert_der_2048,
+            sizeof_client_cert_der_2048);
+        if (ret != 0) {
+            printf("Failed to init with cert\n");
+        }
+    }
+
+    if (pkcs7 != NULL) {
+    #ifdef ECC_TIMING_RESISTANT
+        pkcs7->rng = rng;
+    #endif
+
+        pkcs7->content       = NULL; /* pulling content from callback */
```

Comment:
It doesn't look good to be reaching into structures to set fields.
Are there any public APIs that do any of this?

Reply:
This is kinda the existing setup for PKCS7 use.